src/stdlib/ListUtils.scillib

```diff
@@ -114,82 +176,53 @@ let list_length : forall 'A. List 'A -> Uint32 =
      in
        folder iter init l
 
-(* Helper function for list_eq. Not for public use. *)
-(* Returns Some Nil on successul match. None otherwise. *)
-let list_eq_helper : forall 'A. ('A -> 'A -> Bool) -> 
-  List 'A -> List 'A -> Option (List 'A) =
-  tfun 'A =>
-  fun (eq : 'A -> 'A -> Bool) =>
-  fun (l1 : List 'A) =>
-  fun (l2 : List 'A) =>
-    let folder = @list_foldl 'A (Option (List 'A)) in
-    let init = Some {(List 'A)} l2 in
-    let iter =
-      fun (z : Option (List 'A)) =>
-      fun (h1 : 'A) =>
-        match z with
-        | Some ll2 =>
-          let headF = @list_head 'A in
-          let h2o = headF ll2 in
-          match h2o with
-          | Some h2 =>
-            let eqb = eq h1 h2 in
-            match eqb with
-            | True =>
-              let tailF = @list_tail 'A in
-              tailF ll2
-            | False =>
-              None {(List 'A)}
-            end
-          | None =>
-            None {(List 'A)}
-          end
-        | None =>
-          None {(List 'A)}
-        end
-    in
-      folder iter init l1
-
 (* Return true iff two lists compare equal. *)
 (* Comparison is performed using the "f" function provided. *)
 let list_eq : forall 'A. ('A -> 'A -> Bool) -> List 'A -> List 'A -> Bool =
   tfun 'A =>
-  fun (f : 'A -> 'A -> Bool) =>
-  fun (l1 : List 'A) =>
-  fun (l2 : List 'A) =>
-    let eqh = @list_eq_helper 'A in
-    let res = eqh f l1 l2 in
-    match res with
-    | Some l =>
-      match l with
-      | Nil =>
-        True
-      | _ =>
-        False
-      end
-    | _ =>
-      False
+  fun (equal : 'A -> 'A -> Bool) => fun (l1 : List 'A) => fun (l2 : List 'A) =>
+  (* fold while on non-empty list and fail when not-equal or other empty *)
+  match l1 with
+  | Cons _ _ =>
+    let equal_step = fun (rest: List 'A) => fun (compare: 'A) =>
+      match rest with
+      | Cons x xs => let eq_x_compare = equal x compare in
+        match eq_x_compare with
+        | True => Some {(List 'A)} xs
+        | False => None {(List 'A)}
+        end
+      | Nil => None {(List 'A)}
+      end in
+    let fold_while = @list_foldl_while 'A (List 'A) in
+    let remaining = fold_while equal_step l1 l2 in
+    match remaining with
+    | Cons _ _ => False
+    | Nil => True
+    end
+  (* Special case starting with empty list *)
```

Comment:
Is this special case (and the outer `l1` pattern match) needed at all? It seems to me that it'll work even without this. If `l1` is empty, then `fold_while` will return `l2` as it is and the `Cons` branch will be taken in line 198-199, leading to a false anyway.

Reply:
Indeed, it's not necessary in principle. I don't see an elegant way of removing it if we stick to `fold_while` function (@hgsipiere if you see one, I'd happy to learn your solution). But I see a solution using `list_foldk` function.

Reply:
Removing this change breaks r1 and r4 in the tests but I do agree it is clunky if you thought that.

It is for the case that we pass something list list_eq [] [1,2,3,4] then as our first argument is Nil, we terminate and return back the list [] making them apparently equal. I could try and think of a better way of doing this.